### PR TITLE
Add --path to autoupdate mechanism

### LINKED ISSUE #527: autoupdate does not support --path option
## Problem
`swupd autoupdate` does not accept the `--path=<rootdir>` option like other swupd subcommands: bundle-add, bundle-remove, bundle-list, hashdump, update, verify, check-update, and search. This limits the use of the command to the currently running system.

There is a workaround of using `systemctl` directly, but it would be cleaner to only use the swupd command for consistency and in case systemctl is changed or replaced.

## Use Case
The Clear Linux OS Installer relies on `swupd` for initialization and software installation.

### Enable
`/usr/bin/systemctl --root=<rootdir> unmask --now swupd-update.service swupd-update.timer
`
### Disable
`/usr/bin/systemctl --root=<rootdir> mask --now swupd-update.service swupd-update.timer
`

### PR DESCRIPTION
**Pull request by karthikprabhuvinod**
Signed-off-by: Karthik Prabhu Vinod <karthik.prabhu.vinod@intel.com>

Fixes #527

### CODE REVIEW

**Comment by karthikprabhuvinod** [comment]
@otaviobp : Please could you provide some early feedback

File: src/autoupdate.c [hunk 1]
```diff
@@ -29,55 +29,75 @@
...
-static int disable;
+static bool enable = false;
+static bool disable = false;
 
 static const struct option prog_opts[] = {
 	{ "help", no_argument, 0, 'h' },
-	{ "enable", no_argument, &enable, 1 },
-	{ "disable", no_argument, &disable, 1 },
-	{ 0, 0, 0, 0 }
+	{ "enable", no_argument, 0, 'e' },
```
Comment (otaviobp) [inline src/autoupdate.c]: We don't need short opts for enable and disable. Take a look at bundle-add for an example

File: src/autoupdate.c [hunk 2]
```diff
@@ -29,55 +29,75 @@
...
+
+	if (argc > ind) {
 		error("unexpected arguments\n\n");
 		goto err;
 	}
 
 	return true;
 err:
-	print_help(argv[0]);
+	print_help();
```
Comment (otaviobp) [inline src/autoupdate.c]: don't call print_help() here.

File: src/autoupdate.c [hunk 3]
```diff
@@ -89,13 +109,18 @@ static void policy_warn(void)
 
 enum swupd_code autoupdate_main(int argc, char **argv)
 {
+	char *root_option;
+
 	if (!parse_options(argc, argv)) {
```
Comment (otaviobp) [inline src/autoupdate.c]: Add a print_help() call here, otherwise `swupd autoupdate --invalid` won't print a help message

File: src/lib/sys.h [hunk 4]
```diff
@@ -203,6 +203,12 @@ int sys_rm(const char *filename);
  */
 #define systemctl_cmd(...) run_command_quiet(SYSTEMCTL, __VA_ARGS__)
 
+/**
+ * @brief Run a systemctl command with the informed parameters in a path
+ */
+#define systemctl_cmd_path(path, ...) \
+	path ? run_command_quiet(SYSTEMCTL, path, __VA_ARGS__) : run_command_quiet(SYSTEMCTL, __VA_ARGS__)
```
Comment (otaviobp) [inline src/lib/sys.h]: Instead of receiving a parameter "--root=my_path", we should receive the path here. And call it

    run_command_quiet(SYSTEMCTL, "--root", path, __VA_ARGS__)

It's a better API to have.

Comment (karthikprabhuvinod) [reply src/lib/sys.h]: Would that work... because systemctl man pages expects its as --root=path

Comment (otaviobp) [reply src/lib/sys.h]: Usually `--param=val` is the same as `--param val`

File: src/autoupdate.c (hunk 2 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: src/autoupdate.c (hunk 3 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: src/lib/sys.h (hunk 4 above)
Comment (karthikprabhuvinod) [reply src/lib/sys.h]: Tested works! Thanks i didnt know :)

File: src/autoupdate.c (hunk 1 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: It is good to have short opts isnt it?

Comment (otaviobp) [reply src/autoupdate.c]: We had several problems on conflicts of short opts before. As the number of short opts are limited (we have limited number of letters) we prefer to use it only when needed.

Comment (karthikprabhuvinod) [reply src/autoupdate.c]: ahh I see.

File: test/functional/autoupdate/aup-basic-tests.bats [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+# Author: Karthik Prabhu Vinod
+# Email: karthik.prabhu.vinod@intel.com
+
+load "../testlib"
+
+test_setup() {
+	create_test_environment "$TEST_NAME"
+
+	# Path to install os-install in
+	run sudo sh -c "sudo mkdir '$TEST_DIRNAME'/testroot"
```
Comment (castulo) [inline test/functional/autoupdate/aup-basic-tests.bats]: I don't think you need to create this dir , if you need to install a real bundle into a directory you can install it in "$TARGETDIR" this is supposed to be the target system created by the test environment. However I think this test may be problematic because it is actually installing real bundle so it will need internet connectivity, it is not self contained as the rest are. I understand you need the real bundle so you will have the services there, but that makes me think this test doesn't belong here, maybe belongs to the real content tests https://github.com/clearlinux/swupd-client/tree/master/test/real_content, what do you think @otaviobp

File: Makefile.am [hunk 6]
```diff
@@ -164,6 +164,7 @@ BATS_LOG_DRIVER = $(tap_driver)
 TESTS = $(dist_check_SCRIPTS) $(UNIT_TESTS)
 
 BATS = \
+	test/functional/autoupdate/aup-basic-tests.bats \
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply Makefile.am]: make sense cas..will change

File: src/autoupdate.c [hunk 7]
```diff
@@ -29,56 +29,74 @@
...
 	print("\n");
+
+	global_print_help();
 }
 
-static int enable;
-static int disable;
+#define FLAG_ENABLE 1
+#define FLAG_DISABLE 0
+static bool enable;
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: src/autoupdate.c [hunk 8]
```diff
@@ -105,11 +127,20 @@ enum swupd_code autoupdate_main(int argc, char **argv)
...
 		if (rc != 0) {
 			return SWUPD_SUBPROCESS_ERROR;
 		}
-		rc = systemctl_restart("swupd-update.timer");
+
+		/* If root_path is given and we have succeeded so far
+		 * make sense to return SWUPD_OK at this point
+		 */
+		if (root_path) {
+			return SWUPD_OK;
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: So if we provide root_option, we really cant restart a service because it needs to work with the systemd daemon there. 

So the only thing to be done when we provide root_option is to mask and unmask files.

File: test/functional/README.md [hunk 9]
```diff
@@ -125,6 +125,7 @@ or more of the [assertions](#assertions) provided by the test library.
 
 | Main Business Scenarios                        | Commands      | Group ID |
 | ---------------------------------------------- | ------------- | -------- |
+| Enables/Disables autoupdate                    | auto-update   | AUP      |
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/README.md]: Thanks !!

File: test/functional/autoupdate/aup-basic-tests.bats [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	run sudo sh -c "$SWUPD os-install --bundles=os-core-update --path='$TEST_DIRNAME'/testroot"
+}
+
+@test "AUP001: Basic test, Check auto-update enable" {
+
+	# Initially the file should be all unmasked as autoupdate is enabled
+	assert_file_not_exists "$TEST_DIRNAME'/testroot/etc/systemd/system/swupd-update.service"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic-tests.bats]: Lol yes!! Sorry my bad!

File: test/functional/autoupdate/aup-basic-tests.bats (hunk 5 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic-tests.bats]: Thats a great observation, Thanks for catching that!

File: src/autoupdate.c [hunk 11]
```diff
@@ -29,56 +29,74 @@
 
 #include "swupd.h"
 
-static void print_help(const char *name)
+static void print_help()
 {
-	print("Usage:\n");
-	print("   swupd %s [options]\n\n", basename((char *)name));
 	print("Help Options:\n");
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done ..Thanks!

File: test/functional/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/README.md]: Done

File: test/functional/autoupdate/aup-basic-tests.bats [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+	run sudo sh -c "$SWUPD autoupdate --disable --path='$TEST_DIRNAME'/testroot"
+
+	expected_output=$(cat <<-EOM
+		Warning: disabling automatic updates may take you out of compliance with your IT policy
+		Running systemctl to disable updates
+	EOM
+	)
+
+	assert_status_is 0
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic-tests.bats]: Done

File: test/functional/autoupdate/aup-basic-tests.bats (hunk 10 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic-tests.bats]: Done

File: test/functional/autoupdate/aup-basic-tests.bats [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+	assert_in_output "$expected_output"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic-tests.bats]: Done

**Comment by karthikprabhuvinod** [comment]
@castulo : Thanks for the detailed review.  I have addressed most of the things you asked for and left the others open for @otaviobp to reply on.

File: test/functional/autoupdate/aup-basic-tests.bats (hunk 5 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic-tests.bats]: Looks like I cant even get to pass it on my travis without having systemd daemon installed..I getting this failure: https://travis-ci.org/clearlinux/swupd-client/jobs/596251676#L8185

File: src/autoupdate.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (castulo) [reply src/autoupdate.c]: ok, got it, thanks for the clarification, maybe adding that comment there to explain that in the code for future reference?

File: test/functional/autoupdate/aup-basic-tests.bats (hunk 5 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic-tests.bats]: Done with mocking systemctl ! Thanks for the tip

File: src/autoupdate.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

**Comment by karthikprabhuvinod** [comment]
@otaviobp @castulo : done please review

File: src/autoupdate.c [hunk 14]
```diff
@@ -29,56 +29,76 @@
...
+
+	print("Options:\n");
 	print("       --enable            enable autoupdates\n");
 	print("       --disable           disable autoupdates\n");
 	print("\n");
 }
 
-static int enable;
-static int disable;
+#define FLAG_ENABLE 1
```
Comment (otaviobp) [inline src/autoupdate.c]: Start at 2000 to avoid clashing with any ascii code or global flags

File: src/autoupdate.c [hunk 15]
```diff
@@ -89,47 +109,63 @@ static void policy_warn(void)
...
 	if (!parse_options(argc, argv)) {
+		print_help();
 		return SWUPD_INVALID_OPTION;
 	}
-	if (enable && disable) {
-		error("Can not enable and disable at the same time\n");
-		return SWUPD_INVALID_OPTION;
+
+	if (globals.path_prefix) {
+		string_or_die(&root_path, "%s", globals.path_prefix);
```
Comment (otaviobp) [inline src/autoupdate.c]: You don't need to duplicate memory for path_prefix. Just use the globals.path_prefix where you used root_path.

File: test/functional/autoupdate/aup-basic.bats [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+	create_test_environment "$TEST_NAME"
+
+	run sudo sh -c "mkdir '$TEST_DIRNAME'/fakepath"
+	# mock systemctl if it doesnt exist to always return true
+	if [ ! -e /usr/bin/systemctl ]; then
+		print "systemctl was not found installed in the system"
+		{
+			printf '#!/bin/bash\n'
+			echo ''
```
Comment (otaviobp) [inline test/functional/autoupdate/aup-basic.bats]: echo parameters here so you can check if it has run or not.

File: test/functional/autoupdate/aup-basic.bats [hunk 17]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env bats
+
+# Author: Karthik Prabhu Vinod
+# Email: karthik.prabhu.vinod@intel.com
+
+load "../testlib"
+
+test_setup() {
+
```
Comment (otaviobp) [inline test/functional/autoupdate/aup-basic.bats]: Just run this test on if TRAVIS == true because we write files in /usr and we don't want to do it in dev machines.

File: test/functional/testlib.bash [hunk 18]
```diff
@@ -3169,6 +3169,7 @@ get_next_available_id() { # swupd_function
 	id=$((id+1))
 	test_dir=$(basename "$(realpath "$test_dir")")
 	case "$test_dir" in
+		autoupdate) group=AUP;;
```
Comment (otaviobp) [inline test/functional/testlib.bash]: Why did you change the code?

Comment (karthikprabhuvinod) [reply test/functional/testlib.bash]: It has to three letter. Thats why. Couldnt think of any other

File: test/functional/autoupdate/aup-basic.bats (hunk 17 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Okay

File: src/autoupdate.c (hunk 14 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Okay

File: src/autoupdate.c (hunk 15 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Okay

File: test/functional/autoupdate/aup-basic.bats (hunk 16 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Actually we don't really capture the output of systemctl , so it wont print the parameters.

**Comment by karthikprabhuvinod** [comment]
@otaviobp :So we can turn enabled to disabled, but wont be able to turn disabled to enabled ..for auto-update to be show enabled, swupd-update.timer has to be restarted which we cant for a path

**Comment by otaviobp** [comment]
> @otaviobp :So we can turn enabled to disabled, but wont be able to turn disabled to enabled ..for auto-update to be show enabled, swupd-update.timer has to be restarted which we cant for a path

I think it's safe to ignore the timer if path_prefix != NULL

File: test/functional/testlib.bash (hunk 18 above)
Comment (otaviobp) [reply test/functional/testlib.bash]: Just use what we have defined before AUT. I don't know why @castulo picked that but we already have documentation using that name

File: test/functional/autoupdate/aup-basic.bats (hunk 16 above)
Comment (otaviobp) [reply test/functional/autoupdate/aup-basic.bats]: That's true. But we need to find a way to test if the autoupdate is being turned on or not. This test isn't really testing that.

Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: For that, we will run this on travis where we install systemd like you said and perform os-install(that will require net access)...but then we can perform real tests.

File: test/functional/testlib.bash (hunk 18 above)
Comment (karthikprabhuvinod) [reply test/functional/testlib.bash]: okay

Comment (karthikprabhuvinod) [reply test/functional/testlib.bash]: Done

File: test/functional/autoupdate/aup-basic.bats (hunk 16 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: The we can perform real tests.

File: src/autoupdate.c (hunk 15 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: src/autoupdate.c (hunk 14 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: test/functional/autoupdate/aup-basic.bats [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+	sudo sh -c "ls -l '$PATH_PREFIX'/etc/"
+	sudo sh -c "ls -l '$PATH_PREFIX'/usr/lib"
+
+
+	# perform autoupdate enable
+	run sudo sh -c "$SWUPD autoupdate --enable --path=$PATH_PREFIX"
+	#sudo sh -c "ls -la '$PATH_PREFIX'/etc/systemd/"
+	sudo sh -c "ls -la '$PATH_PREFIX'/usr/lib/systemd/system/"
+	sudo sh -c "ls -la '$PATH_PREFIX'/usr/lib/systemd/system/swupd-update.service"
+	sudo sh -c "ls -la '$PATH_PREFIX'/usr/lib/systemd/system/swupd-update.timer"
```
Comment (otaviobp) [inline test/functional/autoupdate/aup-basic.bats]: /usr/lib/systemd/ is not relevant here. Check the /etc.

Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: removed that was for test

**Comment by karthikprabhuvinod** [comment]
@otaviobp : Thanks for helping find the issue. Please could you review this?

File: src/autoupdate.c [hunk 20]
```diff
@@ -89,54 +109,71 @@ static void policy_warn(void)
...
+		rc2 = systemctl_cmd_path(globals.path_prefix, "is-active", "swupd-update.timer", NULL);
 		if (rc1 == SWUPD_OK && rc2 == SWUPD_OK) {
 			print("Enabled\n");
 			return SWUPD_OK;
 		} else if (rc1 >= STATUS_UNKNOWN || rc2 >= STATUS_UNKNOWN) {
 			/* systemctl returns 1,2, or 3 when program dead or not running */
 			error("Unable to determine autoupdate status\n");
 			return SWUPD_SUBPROCESS_ERROR;
+			// Swupd-service is unmasked, static but timer is inactive for --path
+		} else if (rc1 == SWUPD_OK && rc2 == (STATUS_UNKNOWN - 1) && globals.path_prefix) {
```
Comment (otaviobp) [inline src/autoupdate.c]: Magic number STATUS_UNKNOWN -1. Is this behavior defined in systemctl documentation?

File: src/autoupdate.c [hunk 21]
```diff
@@ -89,54 +109,71 @@ static void policy_warn(void)
...
+			info("swupd-update.service has been unmasked with --path option, but the timer has not been restarted\n");
```
Comment (otaviobp) [inline src/autoupdate.c]: Change print to: Autoupdate is enabled, but the timer is not running

If you run `swupd autoupdate` before running any other command the service hasn't been unmasked (it was never masked). And no one tried to restart the timer.

File: src/autoupdate.c (hunk 20 above)
Comment (otaviobp) [reply src/autoupdate.c]: My suggestion on that is just don't run is-active when path_prefix != "/"

File: test/functional/autoupdate/aup-basic.bats [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+	run sudo sh -c "$SWUPD autoupdate --disable --path=$PATH_PREFIX"
+	assert_status_is "$SWUPD_OK"
+	expected_output=$(cat <<-EOM
+		Warning: disabling automatic updates may take you out of compliance with your IT policy
+		Running systemctl to disable updates
+	EOM
+	)
+	assert_is_output "$expected_output"
+
+	# check for files(should not exist)
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Done

File: test/functional/autoupdate/aup-basic.bats [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+	run sudo sh -c "ls -l '$PATH_PREFIX'/etc/systemd/system/swupd-update.service"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Done

File: test/functional/autoupdate/aup-basic.bats [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+	assert_status_is "$SWUPD_OK"
+	expected_output=$(cat <<-EOM
+		Running systemctl to enable updates
+		Warning: Running autoupdate with --path will not restart swupd-update.timer. This will have to be done manually
+	EOM
+	)
+	assert_is_output "$expected_output"
+
+	# check for files(should not exist)
+	run sudo sh -c "ls -l '$PATH_PREFIX'/etc/systemd/system/swupd-update.service"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Done

File: .travis.yml [hunk 25]
```diff
@@ -51,6 +51,8 @@ install:
         - sudo apt-get install doxygen
         - sudo apt-get install check
         - sudo pip install coverxygen
+        - sudo apt-get install systemd
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply .travis.yml]: Okay will remove this

File: src/autoupdate.c (hunk 21 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: You are right..changed it

File: src/autoupdate.c (hunk 20 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: I can do that..problem is the globals.path_prefix has no default value of '/' value in it as we dont perform SWUPD_INIT ad SWUPD_DEINIT in case of autoupdate as it is not really required. Should we?

Comment (karthikprabhuvinod) [reply src/autoupdate.c]: You rightly pointed out ..there is no documentation for isactive for Magic number STATUS_UNKNOWN -1. The only thing they say it returns a number other than zero for in-active.

File: src/autoupdate.c [hunk 26]
```diff
@@ -89,54 +109,71 @@ static void policy_warn(void)
...
+		rc2 = systemctl_cmd_path(globals.path_prefix, "is-active", "swupd-update.timer", NULL);
 		if (rc1 == SWUPD_OK && rc2 == SWUPD_OK) {
 			print("Enabled\n");
 			return SWUPD_OK;
 		} else if (rc1 >= STATUS_UNKNOWN || rc2 >= STATUS_UNKNOWN) {
 			/* systemctl returns 1,2, or 3 when program dead or not running */
 			error("Unable to determine autoupdate status\n");
 			return SWUPD_SUBPROCESS_ERROR;
+			// Swupd-service is unmasked, static but timer is inactive for --path
+		} else if (rc1 == SWUPD_OK && rc2 != SWUPD_OK && globals.path_prefix) {
```
Comment (otaviobp) [inline src/autoupdate.c]: good fix

File: .travis.yml [hunk 27]
```diff
@@ -12,7 +12,7 @@ env:
                 - UPDATE_TOTAL="$(find test/functional/update -name *.bats | wc -l)"
                 - UPDATE_SUBGROUP1_TOTAL=17
                 - UPDATE_SUBGROUP2_TOTAL="$((UPDATE_TOTAL - UPDATE_SUBGROUP1_TOTAL))"
-                - UPDATE_SUBGROUP1="$(find test/functional/update -name *.bats | head -n $UPDATE_SUBGROUP1_TOTAL | tr '\n' ' ')"
+                - UPDATE_SUBGROUP1="$(find test/functional/{autoupdate, update} -name *.bats | head -n $UPDATE_SUBGROUP1_TOTAL | tr '\n' ' ')"
```
Comment (castulo) [inline .travis.yml]: don't add these here, you need to move this in a group not a subgroup

File: src/autoupdate.c [hunk 28]
```diff
@@ -29,56 +29,76 @@
...
-		case 0: /* getopt_long has set the flag */
-			break;
-		default:
-			error("unrecognized option\n\n");
-			goto err;
+	switch (opt) {
+	case 'h':
+		print_help();
+		exit(EXIT_SUCCESS);
+	case FLAG_ENABLE:
```
Comment (castulo) [inline src/autoupdate.c]: I told you this in a previous command but it is better to check flag restrictions within parse_options() as we do in the other commands

File: test/functional/autoupdate/aup-basic.bats [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+	create_test_environment "$TEST_NAME"
+
+}
+
+test_teardown() {
+
+	if [ -z "$TRAVIS" ]; then
+		return
+	fi
+
```
Comment (castulo) [inline test/functional/autoupdate/aup-basic.bats]: you are never cleaning the tea environment , if I run the test with TRAVIS=true the test env will be left behind

File: test/functional/autoupdate/aup-basic.bats [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+	assert_status_is "$SWUPD_OK"
+	expected_output=$(cat <<-EOM
+		Warning: disabling automatic updates may take you out of compliance with your IT policy
+		Running systemctl to disable updates
+	EOM
+	)
+	assert_is_output "$expected_output"
+
+	# check for files(should exist)
+	assert_file_exists "'$PATH_PREFIX'/etc/systemd/system/swupd-update.service"
```
Comment (castulo) [inline test/functional/autoupdate/aup-basic.bats]: why the single quote?

File: test/functional/autoupdate/aup-basic.bats [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+	if [ -z "$TRAVIS" ]; then
+		return
+	fi
+
+}
+
+@test "AUT001: Basic test, Check auto-update enable" {
+
+	# perform autoupdate disable
+	run sudo sh -c "$SWUPD autoupdate --disable --path=$PATH_PREFIX"
```
Comment (castulo) [inline test/functional/autoupdate/aup-basic.bats]: where are you defining PATH_PREFIX?

File: .travis.yml (hunk 27 above)
Comment (castulo) [reply .travis.yml]: if you check your Travis execution you will notice about 25 tests were not run because of this change

Comment (karthikprabhuvinod) [reply .travis.yml]: Yikes, Yes..I have an extra space. My bad

File: test/functional/autoupdate/aup-basic.bats (hunk 29 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Ahh Yes..great catch..neeed to add a destroy

File: test/functional/autoupdate/aup-basic.bats (hunk 30 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: I thought I  needed it .. doesnt look like. Thanks.

File: test/functional/autoupdate/aup-basic.bats (hunk 31 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: This is created as part of the test environment ..testlib

File: .travis.yml (hunk 27 above)
Comment (karthikprabhuvinod) [reply .travis.yml]: Done

File: test/functional/autoupdate/aup-basic.bats (hunk 29 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Done

File: test/functional/autoupdate/aup-basic.bats (hunk 30 above)
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: Done

File: test/functional/autoupdate/aup-basic.bats (hunk 31 above)
Comment (castulo) [reply test/functional/autoupdate/aup-basic.bats]: right I don't know how I missed that, I see it know, thanks

File: .travis.yml (hunk 27 above)
Comment (castulo) [reply .travis.yml]: please don't add autoupdate to UPDATE_SUBGROUP1 that is ment for update related tests, you want the run as part of worker 1 so add autoupdate to GROUP1

File: src/autoupdate.c (hunk 28 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: test/functional/autoupdate/aup-basic.bats (hunk 24 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: @otaviobp I keep forgetting this, assert_file will not work in this case because the file is symlinked ...I think assert_file does not check for this

File: .travis.yml (hunk 27 above)
Comment (karthikprabhuvinod) [reply .travis.yml]: Yes corrected it

File: src/autoupdate.c (hunk 20 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: test/functional/autoupdate/aup-basic.bats (hunk 24 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply test/functional/autoupdate/aup-basic.bats]: @castulo and I discussed..looks like there is a test -c function to test for char file..didnt want to change function assert_file_exists

File: src/autoupdate.c [hunk 32]
```diff
@@ -29,56 +29,76 @@
...
+};
+
+static bool parse_opt(int opt, UNUSED_PARAM char *optarg)
+{
+	switch (opt) {
+	case 'h':
+		print_help();
+		exit(EXIT_SUCCESS);
+	case FLAG_ENABLE:
+		cmdline_option_enable = true;
```
Comment (castulo) [inline src/autoupdate.c]: ups, one more thing, instead of directly assigning `true` you need to use the `optarg_to_bool` macro like this `cmdline_option_enable = optarg_to_bool(optarg);` otherwise if those values are added to the config file they won't be parsed correctly.

File: src/autoupdate.c [hunk 33]
```diff
@@ -29,56 +29,76 @@
...
 	}
 
-	if (argc > optind) {
+	if (argc > ind) {
 		error("unexpected arguments\n\n");
-		goto err;
+		return false;
+	}
+
+	if (cmdline_option_enable) {
```
Comment (castulo) [inline src/autoupdate.c]: thanks for making the change. In this case since there are only two conflicting flags I think is better to use just one if statement like:
 `if (cmdline_option_enable && cmdline_option_disable) {`

we can use nested ifs when there are multiple conflicting flags

Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Done

File: src/autoupdate.c (hunk 32 above)
Comment (karthikprabhuvinod) [reply src/autoupdate.c]: Good catch! Thanks Done..
